tests/loan_protocol.rs

```diff
@@ -497,6 +515,156 @@ async fn lend_and_dynamic_liquidate() {
     }
 }
 
+#[tokio::test]
+async fn can_run_protocol_with_principal_change_outputs() {
+    let mut rng = ChaChaRng::seed_from_u64(0);
+
+    let bitcoin_asset_id = "0c5d451941f37b801d04c46920f2bc5bbd3986e5f56cb56c6b17bedc655e9fc6"
+        .parse()
+        .unwrap();
+    let usdt_asset_id = "6b397062b69411b554ec398ae3b25fdc54fab1805126786581a56a7746afbab2"
+        .parse()
+        .unwrap();
+    let (_oracle_sk, oracle_pk) = make_keypair(&mut rng);
+
+    let mut wallet = Wallet::default();
+    let borrower = {
+        let (_sk, pk) = make_keypair(&mut rng);
+        let (blinding_sk, blinding_pk) = make_keypair(&mut rng);
+        let address = Address::p2pkh(&pk, Some(blinding_pk.key), &AddressParams::LIQUID);
+
+        let collateral_amount = Amount::ONE_BTC;
+
+        Borrower0::new(
+            &mut rng,
+            {
+                let wallet = &mut wallet;
+                |amount, asset| async move {
+                    wallet.coin_select(
+                        // we need extra coins to pay for the fee
+                        amount + Amount::from_sat(10_000),
+                        asset,
+                    )
+                }
+            },
```

Comment:
This doesn't work?

```suggestion
|amount, asset| async { 
	wallet.coin_select(
        // we need extra coins to pay for the fee
        amount + Amount::from_sat(10_000),
        asset,
    )
 },
```

Reply:
I'd bet money on having tried that, but I'll try again :stuck_out_tongue:

Reply:
The thing is:

1. unless there are lifetime restrictions, `move` should not be necessary.

2. if there are lifetime restrictions through bounds on the argument, we should remove them because they should be unnecessary.

`move` is usually only necessary for spawning into an executor. Those APIs have a 'static bound because they need to be able to move stuff between (executor) threads.

We don't do that here but just await the future internally so the requirement for `move` should fall away.

Really, you want to get rid of as many `move`s as possible. It is one of those things in Rust that have a ripple effect in complexity.

Reply:
I think I get your points but I don't know how to solve it:

```
error[E0373]: async block may outlive the current function, but it borrows `amount`, which is owned by the current function
   --> tests/loan_protocol.rs:541:39
    |
541 |                   |amount, asset| async {
    |  _______________________________________^
542 | |                     wallet.coin_select(
543 | |                         // we need extra coins to pay for the fee
544 | |                         amount + Amount::from_sat(10_000),
    | |                         ------ `amount` is borrowed here
545 | |                         asset,
546 | |                     )
547 | |                 }
    | |_________________^ may outlive borrowed value `amount`
    |
note: async block is returned here
   --> tests/loan_protocol.rs:541:33
    |
541 |                   |amount, asset| async {
    |  _________________________________^
542 | |                     wallet.coin_select(
543 | |                         // we need extra coins to pay for the fee
544 | |                         amount + Amount::from_sat(10_000),
545 | |                         asset,
546 | |                     )
547 | |                 }
    | |_________________^
help: to force the async block to take ownership of `amount` (and any other referenced variables), use the `move` keyword
    |
541 |                 |amount, asset| async move {
542 |                     wallet.coin_select(
543 |                         // we need extra coins to pay for the fee
544 |                         amount + Amount::from_sat(10_000),
545 |                         asset,
546 |                     )
```

Reply:
Interestingly this doesn't happen when I pass in a closure to sign the inputs:

```rust
 let loan_transaction = lender
        .finalise_loan(loan_transaction, {
            |tx| async { Ok(wallet.sign_inputs(tx)) }
        })
        .await
        .unwrap();
```

The difference that I notice is that the problematic one takes a mutable reference to wallet and this last one an immutable one. But I don't know why that would affect the arguments of the closure.

Reply:
> Well, this should be simpler in the future by not having these signer closures anyway :)

That's what I think too!